go/marshal/decode.go

```diff
@@ -217,3 +229,75 @@ func nomsValueDecoder(v types.Value, rv reflect.Value) {
 	}
 	rv.Set(reflect.ValueOf(v))
 }
+
+func sliceDecoder(t reflect.Type) decoderFunc {
+	decoderCache.RLock()
+	d := decoderCache.m[t]
+	decoderCache.RUnlock()
+	if d != nil {
+		return d
+	}
+
+	var decoder decoderFunc
+
+	d = func(v types.Value, rv reflect.Value) {
+		list := v.(types.List)
+		if list.Len() == 0 {
+			rv.Set(reflect.MakeSlice(t, 0, 0))
+			return
+		}
+
+		slice := rv.Slice(0, 0)
+		list.IterAll(func(v types.Value, i uint64) {
+			elemRv := reflect.New(t.Elem()).Elem()
+			decoder(v, elemRv)
+			slice = reflect.Append(slice, elemRv)
+		})
+		rv.Set(slice)
+	}
+
+	decoderCache.Lock()
+	if decoderCache.m == nil {
+		decoderCache.m = map[reflect.Type]decoderFunc{}
+	}
+	decoderCache.m[t] = d
+	decoderCache.Unlock()
+
+	decoder = typeDecoder(t.Elem())
+
+	return d
+}
+
+func arrayDecoder(t reflect.Type) decoderFunc {
+	decoderCache.RLock()
+	d := decoderCache.m[t]
+	decoderCache.RUnlock()
+	if d != nil {
+		return d
+	}
```

Comment:
yeah, wonder if you should be able to go
```
d := decoderCache.Get(t)
// and
decoderCache.Set(t, d) // for below
```
save a lot of the lock/unlock repetition.